src/data/sparse_page_dmatrix.cc

```diff
@@ -12,261 +12,91 @@
+  void operator++() override { at_end_ = !source_->Next(); }
+  bool AtEnd() const override { return at_end_; }
+  SparseBatchIteratorImpl* Clone() override {
+    return new SparseBatchIteratorImpl(*this);
   }
-}
-
-void SparsePageDMatrix::ColPageIter::Init(
-    const std::vector<bst_uint>& index_set) {
-  set_index_set_ = index_set;
-  set_load_all_ = true;
-  std::sort(set_index_set_.begin(), set_index_set_.end());
-  this->BeforeFirst();
-}
 
-  dmlc::DataIter<SparsePage>* SparsePageDMatrix::ColIterator() {
-  CHECK(col_iter_ != nullptr);
-  std::vector<bst_uint> col_index;
-  std::iota(col_index.begin(), col_index.end(), bst_uint(0));
-  col_iter_->Init(col_index);
-  return col_iter_.get();
+ private:
+  SparsePageSource* source_{nullptr};
+  bool at_end_{ false };
+};
+
+BatchSet SparsePageDMatrix::GetRowBatches() {
+  auto cast = dynamic_cast<SparsePageSource*>(row_source_.get());
+  cast->BeforeFirst();
+  cast->Next();
+  auto begin_iter = BatchIterator(new SparseBatchIteratorImpl(cast));
+  return BatchSet(begin_iter);
 }
 
-bool SparsePageDMatrix::TryInitColData(bool sorted) {
-  // load meta data.
-  std::vector<std::string> cache_shards = common::Split(cache_info_, ':');
-  {
-    std::string col_meta_name = cache_shards[0] + ".col.meta";
-    std::unique_ptr<dmlc::Stream> fmeta(
-        dmlc::Stream::Create(col_meta_name.c_str(), "r", true));
-    if (fmeta == nullptr) return false;
-    CHECK(fmeta->Read(&buffered_rowset_)) << "invalid col.meta file";
-    CHECK(fmeta->Read(&col_size_)) << "invalid col.meta file";
-  }
-  // load real data
-  std::vector<std::unique_ptr<dmlc::SeekStream> > files;
-  for (const std::string& prefix : cache_shards) {
-    std::string col_data_name = prefix + ".col.page";
-    std::unique_ptr<dmlc::SeekStream> fdata(
-        dmlc::SeekStream::CreateForRead(col_data_name.c_str(), true));
-    if (fdata == nullptr) return false;
-    files.push_back(std::move(fdata));
+BatchSet SparsePageDMatrix::GetSortedColumnBatches() {
+  // Lazily instantiate
+  if (!sorted_column_source_) {
+    SparsePageSource::CreateColumnPage(this, cache_info_, true);
+    sorted_column_source_.reset(
+        new SparsePageSource(cache_info_, ".sorted.col.page"));
   }
-  col_iter_.reset(new ColPageIter(std::move(files)));
-  // warning: no attempt to check here whether the cached data was sorted
-  col_iter_->sorted = sorted;
-  return true;
+  sorted_column_source_->BeforeFirst();
+  sorted_column_source_->Next();
+  auto begin_iter =
+      BatchIterator(new SparseBatchIteratorImpl(sorted_column_source_.get()));
+  return BatchSet(begin_iter);
 }
 
-void SparsePageDMatrix::InitColAccess(
-  size_t max_row_perbatch, bool sorted) {
-  if (HaveColAccess(sorted)) return;
-  if (TryInitColData(sorted)) return;
-  const MetaInfo& info = this->Info();
-  if (max_row_perbatch == std::numeric_limits<size_t>::max()) {
-    max_row_perbatch = kMaxRowPerBatch;
```

Comment:
I notice that there is no reference to `max_row_perbatch` or `kMaxRowPerBatch` anywhere in your codebase. Are we exclusively using `kPageSize` to determine page sizes? If so, we should remove `max_row_perbatch` parameter and `kMaxRowPerBatch`  constant.

Reply:
Yes, I will do this.